src/esw/auton_led.py

```diff
@@ -2,82 +2,123 @@
-            green_counter_s = 0
-            ser.write(b"g")
-        elif color == "blue":
-            ser.write(b"b")
-        else:
-            ser.write(b"o")
+        self._green_counter_s = 0
 
-        previous_color = color
+        # create serial connection with Arduino
+        self._ser = serial.Serial(port=port, baudrate=baud)
 
-    return ChangeAutonLEDStateResponse(True)
+        self.update()
 
+    def handle_change_state(self, req: ChangeAutonLEDStateRequest) -> ChangeAutonLEDStateResponse:
+        """
+        Processes a request to change the auton LED array state by changing the desired color.
+        Returns the success of the transaction.
 
-def main():
-    rospy.init_node("auton_led")
+        :param req: A string that is the color of the requested state of the
+            auton LED array. Note that green actually means blinking green.
+        :returns: A boolean that is always True.
+        """
 
-    global ser, previous_color, green_counter_s
+        with self._color_lock:
+            self._color = req.color.lower()
 
-    # read serial connection info from parameter server
-    port = rospy.get_param("auton_led_driver/port")
-    baud = rospy.get_param("auton_led_driver/baud")
+        self.update()
 
-    # create serial connection with Arduino
-    ser = serial.Serial(port, baud)
+        return ChangeAutonLEDStateResponse(True)
 
-    rospy.Service("change_auton_led_state", ChangeAutonLEDState, handle_change_auton_led_state)
+    def flash_if_green(self):
+        """
+        Updates serial and green counter if the requested color is green.
+        """
+        with self._color_lock:
 
-    seconds_to_wait = 1
+            if self._color != "green":
+                return
 
-    while not rospy.is_shutdown():
-        if previous_color == "green":
+            # If requested color is green, then alternate between off and on.
+            if self._green_counter_s >= self.GREEN_PERIOD_S:
+                self._green_counter_s = 0
+
+            if self._green_counter_s == 0:
+                self._ser.write(b"g")
+            elif self._green_counter_s >= self.GREEN_ON_S:
+                self._ser.write(b"o")
+
+    def sleep(self):
+        """
+        Sleeps and updates _green_counter_s if necessary.
+        """
+        time.sleep(self.SLEEP_AMOUNT)
+
+        with self._color_lock:
             # if requested color is green,
-            # then alternate between off and on
-            # every second to make it a blinking LED
+            if self._color == "green":
+                self._green_counter_s += self.SLEEP_AMOUNT
+
+    def update(self):
+        """
+        Writes to serial to change LED color.
+        """
+        with self._color_lock:
+            if self._color == "red":
+                self._ser.write(b"r")
 
-            if green_counter_s == 2:
-                green_counter_s = 0
+            elif self._color == "green":
+                self._ser.write(b"g")
```

Comment:
You could use a dictionary here if you want and maybe it'll be cleaner but like I think that might be a preference thing maybe it's not cleaner lol up to you guys and I understand just wanting to merge

Reply:
This has not been tested yet, so should not be merged before testing (continue to use master until then). But yeah I think that's a good idea, will push in a little bit